File: Difficulty: Hard/Nodes at given distance in binary tree/nodes-at-given-distance-in-binary-tree.cpp
/* A binary Tree node
struct Node
{
    int data;
    struct Node *left, *right;
};
*/

class Solution {

  public:
  
    void  markparents(Node* root,unordered_map<Node* ,Node* > &parent_track,Node* target_node){
        queue<Node* > q;
        q.push(root);
        while(!q.empty()){
            Node* curr = q.front();
            q.pop();
            if(curr->left){
                parent_track[curr->left] = curr;
                q.push(curr->left);
            }
            if(curr->right){
                parent_track[curr->right] = curr;
                q.push(curr->right);
            }
            
        }
    }
  
     Node* findTarget(Node* root, int target) {
        if (root == NULL) {
            return NULL;
        }
        
        if (root->data == target) {
            return root;
        }
        
        Node* leftResult = findTarget(root->left, target);
        if (leftResult != NULL) {
            return leftResult;
        }
        
        Node* rightResult = findTarget(root->right, target);
        return rightResult;
    }
    vector<int> KDistanceNodes(Node* root, int target, int k) {
        // return the sorted vector of all nodes at k dist
        Node* target_node = findTarget(root,target);
        unordered_map<Node*,Node*> parent_track;
        markparents(root,parent_track,target_node);
        
        unordered_map<Node* ,bool> vis;
        queue<Node*> q;
        q.push(target_node);
        vis[target_node]=true;
        int curr_level=0;
        while(!q.empty()){
            int size = q.size();
            if(curr_level++ == k) break;
            for(int i=0;i<size;i++){
                Node* curr = q.front();
                q.pop();
                if(curr->left && !vis[curr->left]){
                    q.push(curr->left);
                    vis[curr->left] = true;
                }
                if(curr->right && !vis[curr->right]){
                    q.push(curr->right);
                    vis[curr->right] = true;
                }
                if(parent_track[curr] && !vis[parent_track[curr]]){
                    q.push(parent_track[curr]);
                    vis[parent_track[curr]] = true;
                }
            }
        }
        vector<int> res;
        while(!q.empty()){
            Node* curr = q.front();
            q.pop();
            res.push_back(curr->data);
        }
        sort(res.begin(), res.end());

        return res;
        
    }
    

        
};

File: Difficulty: Hard/Nodes at given distance in binary tree/nodes-at-given-distance-in-binary-tree.py
'''
# Tree Node
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
from collections import deque, defaultdict
class Solution:
    
    def KDistanceNodes(self,root,target,k):
        # code here
        # return the sorted list all nodes at k distance from target
        
        res=[]
        if root is None:
            return res
        
        q=deque([root])
        mpp= {root:None}
        tar=0
        
        while q:
            curr= q.popleft()
            
            if curr.data==target:
                tar=curr
            
            if curr.left:
                mpp[curr.left]=curr
                q.append(curr.left)
            if curr.right:
                mpp[curr.right]=curr
                q.append(curr.right)
                
        q.append(tar)
        
        vis={}
        while q:
            size=len(q)
            
            if k==0:
                while q:
                    curr=q.popleft()
                    res.append(curr.data)
                break
            
            while size>0:
                node = q.popleft()
                vis[node]=True
                
                if node.left and not vis.get(node.left,False):
                    q.append(node.left)
                if node.right and not vis.get(node.right,False):
                    q.append(node.right)
                
                parent = mpp.get(node)
                if parent and not vis.get(parent,False):
                    q.append(parent)
                
                size-=1
            k-=1
        
        res.sort()
        return res
                
                
            
            